nucliadb/nucliadb/search/search/chat/prompt.py

```diff
@@ -365,11 +376,34 @@ async def build(self) -> tuple[PromptContext, PromptContextOrder]:
                 f"Prompt context size exceeded: {ccontext.size}."
                 f"The context will be truncated to the maximum size: {self.max_context_size}."
             )
+
+        if self.visual_llm:
+            await self._build_context_images(ccontext)
+
         context = ccontext.output
+        context_images = ccontext.images
         context_order = {
             text_block_id: order for order, text_block_id in enumerate(context.keys())
         }
-        return context, context_order
+        return context, context_order, context_images
+
+    async def _build_context_images(self, context: CappedPromptContext) -> None:
+        ordered_paras = get_ordered_paragraphs(self.find_results)
+        for paragraph in ordered_paras:
+            if paragraph.page_with_visual and paragraph.position:
+                if (
+                    self.image_strategies == ImageRagStrategyName.PAGE_IMAGE
+                    and paragraph.position.page_number
+                ):
+                    page = paragraph.position.page_number
+                    context.images[paragraph.id] = await get_page_image(
+                        self.kbid, paragraph.id, page
+                    )
+                elif paragraph.reference is not None:
+                    image = paragraph.reference
+                    context.images[paragraph.id] = await get_paragraph_image(
+                        self.kbid, paragraph.id, image
+                    )
```

Comment:
Could it be that the same image is referenced in more than one paragraph? If so, we would be downloading it twice.

Also, images are serially downloaded right now, which can be slow at query time. Should we parallelize it?